pkg/eventingtls/eventingtls.go

```diff
@@ -0,0 +1,200 @@
+
+	certHolder := atomic.Value{}
+
+	logger := logging.FromContext(ctx).Desugar().
+		With(zap.String("tls.secret", secret.String()))
+
+	store := func(obj interface{}) {
+		s, ok := obj.(*corev1.Secret)
+		if !ok {
+			return
+		}
+		crt, crtOk := s.Data[TLSCrt]
+		key, keyOk := s.Data[TLSKey]
+		if !crtOk || !keyOk {
+			logger.Debug("Missing " + TLSCrt + " or " + TLSKey + " in the secret.data")
+			return
+		}
+
+		logger.Debug("Loading key pair")
+
+		certificate, err := tls.X509KeyPair(crt, key)
+		if err != nil {
+			return
+		}
+
+		logger.Debug("certificate stored")
+		certHolder.Store(&certificate)
+	}
+
+	informer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controller.FilterWithNameAndNamespace(secret.Namespace, secret.Name),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: store,
+			UpdateFunc: func(_, newObj interface{}) {
+				store(newObj)
+			},
+			DeleteFunc: nil,
+		},
+	})
+
+	// Store the current value so that we have certHolder initialized.
+	firstValue, err := informer.Lister().Secrets(secret.Namespace).Get(secret.Name)
+	if err != nil {
+		// Try to get the secret from the API Server when the lister failed.
+		firstValue, err = kube.CoreV1().Secrets(secret.Namespace).Get(ctx, secret.Name, metav1.GetOptions{})
+		if err != nil {
+			logger.Fatal(err.Error())
+		}
+	}
+	store(firstValue)
+
+	return func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		cert := certHolder.Load()
+		if cert == nil {
+			return nil, nil
+		}
+		return cert.(*tls.Certificate), nil
+	}
+}
+
+// NewDefaultClientConfig returns a default ClientConfig.
+func NewDefaultClientConfig() ClientConfig {
+	return ClientConfig{}
+}
+
+// GetTLSClientConfig returns tls.Config based on the given ClientConfig.
+func GetTLSClientConfig(config ClientConfig) (*tls.Config, error) {
+	pool, err := certPool(config.CACerts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tls.Config{
+		RootCAs:    pool,
+		MinVersion: DefaultMinTLSVersion,
+	}, nil
+}
+
+func NewDefaultServerConfig() ServerConfig {
+	return ServerConfig{}
+}
+
+func GetTLSServerConfig(config ServerConfig) (*tls.Config, error) {
+	return &tls.Config{
+		MinVersion:     DefaultMinTLSVersion,
+		GetCertificate: config.GetCertificate,
+	}, nil
+}
+
+// IsHttpsSink returns true if the sink has scheme equal to https.
+func IsHttpsSink(sink string) bool {
+	s, err := apis.ParseURL(sink)
+	if err != nil {
+		return false
+	}
+	return strings.EqualFold(s.Scheme, "https")
+}
+
+// certPool returns a x509.CertPool with the combined certs from:
+// - the system cert pool
+// - the given CA certificates
+func certPool(caCerts *string) (*x509.CertPool, error) {
+	p, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, err
+	}
+
+	// Clone the pool before appending other certs or returning since we don't want to add user
+	// provided CA certs for different resources to the system pool or to any other "global" pool.
+	p = p.Clone()
```

Comment:
I think it's safe to assume it'll already be a copy. From `SystemCertPool `:

```
// SystemCertPool returns a copy of the system cert pool.
// Any mutations to the returned pool are not written to disk and do not affect
// any other pool returned by SystemCertPool.
```

Reply:
Removed